File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        def func(goal):
            if goal < 0:
                return 0
            l = 0
            r = 0
            curr = 0
            count = 0
            while(r < len(nums)):
                curr+=nums[r]
                while(curr > goal):
                    curr-=nums[l]
                    l+=1
                count+=(r - l + 1)
                r+=1
            return count
        return func(goal) - func(goal-1)

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int k) {
        HashMap<Integer,Integer> map = new HashMap<>();
        map.put(0,1);
        int sum = 0;
        int count = 0;
        for(int i = 0;i<nums.length;i++){
            sum+=nums[i];
            count+=map.getOrDefault(sum - k,0);
            map.put(sum, map.getOrDefault(sum,0) + 1);
        }
        return count;
    }
}